### ISSUE TITLE
Redirects cause root route loader invocation but root route component does not receive loader data

### ISSUE THREAD

**Issue by jangerhofer** [remix-run/remix#3302]
### What version of Remix are you using?

1.5.1

### Steps to Reproduce

A repo with a minimal reproduction is available here: https://github.com/jangerhofer/remixRootRouteRedirectReproduction along with some directions.

### Expected Behavior

I expect that a server-side redirect (whether performance only once or multiple times) would cause a fresh root route `loader` function invocation and that the (serialized) value returned from the `loader` would be made available to the root route component via the `useLoaderData` hook.

### Actual Behavior

Instead, I see only the invocation the root route `loader`; the `useLoaderData` hook's value remains an empty object despite the `loader` returning data.

**Comment by amanolar94** [comment]
@jangerhofer did you maybe have any luck with this? running onto the same problem as well...

**Comment by jangerhofer** [comment]
@amanolar94: I did not, unfortunately.

Instead, I moved around my route structure to avoid having to perform two redirects. I used a [pathless layout route](https://remix.run/docs/en/v1/guides/routing#pathless-layout-routes) in the process.

**Comment by amanolar94** [comment]
I see..In our case pathless layout route won't do since it is a [dynamic segment](https://remix.run/docs/en/v1/guides/routing#dynamic-segments).

**Comment by jamesilva** [comment]
@jangerhofer I was having a similar bug, but in the opposite way, where a user that no longer was logged in was still showing up in the root navbar. Then I started digging and found this  post, and in your repo you say that it works if

> "the login form redirects to /, but the / index route's loader does not redirect but rather renders a component itself"

so I tried your code and only changed a few things in the index.tsx, leaving the routing to the component:
```
export const loader: LoaderFunction = async ({ request }) => {
  const cookieHeader = request.headers.get('Cookie');
  const { email } = (await session.parse(cookieHeader)) || {};

  return { email };
};

export default function Index() {
  let { email } = useLoaderData();
  let navigate = useNavigate();

  useEffect(() => {
    if (email) {
      navigate('/content');
    } else {
      navigate('login');
    }
  }, [email]);
  return <></>;
}
```
which worked as intended. 
So, after a bit more digging, I finally got to the remix page about resource routes [https://remix.run/docs/en/v1/guides/resource-routes](url), which says:

> If a route doesn't export a default component, it can be used as a Resource Route. If called with GET, the loader's response is returned and none of the parent route loaders are called either (because those are needed for the UI, but this is not the UI). If called with POST, the action's response is called.

Now, this could be an answer, if if wasn't for the fact that we know the root loader is running, but its the root `useLoaderData()` which isn't (?) So, I'm not sure if that's what the documentation means and that's what is indeed happening or if it's something entirely different.

Also, `redirect` is just an http wrapper that creates a new `Response` with a redirect status, so it would make sense that we only return it in a `loader` when it is in a resource route. Otherwise, it would be a `Response` to be read when the client is expecting data. The only time I see `redirect` being used in the loading stage is by the server throwing it for the client to handle, otherwise it's probably only used in `actions`.

It's my first time answering to a question, and I'm new to remix, so I'm sorry if it's a bit confusing but it was something that also puzzled me and I think I kinda got deeper and had some info that might help!

**Comment by ryanflorence** [comment]
Definitely sounds like multi-redirect is the problem here. I'm pretty sure our new code over in React Router doesn't do this. @brophdawg11 you mind looking into this one, fixing in Remix and verifying we have a test for it in React Router?

Should be able to make a test case out of the repo² pretty easily (get it? reproduction repository, the repo squared?! ... yeah it's early).

**Comment by ryanflorence** [comment]
What's happening is Remix is accidentally trying to optimize the transition by only calling the loaders of changing layouts. When an action is called, it's not supposed to do that (it calls all loaders) but the double redirect has a bug (it seems) that still tries to optimize and so the root loader isn't called.

You can work around this bug for now by using `<Form reloadDocument>` on the login page so that everything runs through full document requests (and therefore skips any of the client side optimization attempts).

Or you can add some junk to the url search params that'll trigger all loaders to be called:

```jsx
redirect("/content?"+Date.now());
```

**Comment by MichaelDeBoey** [comment]
Closed by #3460
